lib/astutils.cpp

```diff
@@ -277,7 +277,8 @@ static bool hasToken(const Token * startTok, const Token * stopTok, const Token
     return false;
 }
 
-const Token * nextAfterAstRightmostLeaf(const Token * tok)
+template <class T>
+T* nextAfterAstRightmostLeafGeneric(T* tok)
```

Comment:
> personally I would rather use const_cast than template. 

The problem with `const_cast` is you can silently invoke UB without a warning from the compiler, whereas with the template you get a nice clear message about the error, like this:

```
cppcheck/lib/astutils.cpp:283:5: error: member function 'swapWithNext' not viable: 'this' argument has type 'const Token', but function is not marked const
    tok->swapWithNext();
    ^~~
cppcheck/lib/astutils.cpp:300:67: note: in instantiation of function template specialization 'nextAfterAstRightmostLeafGeneric<const Token>' requested here
const Token* nextAfterAstRightmostLeaf(const Token* tok) { return nextAfterAstRightmostLeafGeneric(tok); }
                                                                  ^
cppcheck/lib/token.h:218:10: note: 'swapWithNext' declared here
    void swapWithNext();
         ^
```

> Templates are "raw" code, like macros.

No they are not. Templates are checked for syntax and partially for semantics(only sema is delayed for types dependent on the template type).

> The compiler error messages can be really cryptic.

The error should be the same, just with an extra note about how and where it was instantiated.

Usually cryptic errors come from templates with implicit constraints that are not checked(by `enable_if` or `static_assert`), and in this case the errors might get confusing if we called the function with `Scope*` or some other type that isn't a `Token`.

However, I thought it was kind of obvious from the code that it needed a `Token` class, but I could add a `static_assert` if you think it can be confusing.

Reply:
> However, I thought it was kind of obvious from the code that it needed a Token class, but I could add a static_assert if you think it can be confusing.

If you can improve the type safety and we could get sensible errors in case of mistake, I think that sounds good.

Reply:
> whereas with the template you get a nice clear message about the error, like this:

I don't agree that the message is clear. It would be a clear message if it pointed out the instantiation and said what the given type is and what the expected type is. The stuff about the template implementation is noise.

Reply:
> It would be a clear message if it pointed out the instantiation and said what the given type is and what the expected type is. The stuff about the template implementation is noise.

The method is given the expected type. What is incorrect is calling a non-const member function on the variable(I inserted a call to `swapWithNext` to see what the message would look like). The compiler catches this and points to the incorrect method.

Using `const_cast` this would not be caught. We could add a check in the future to try to catch, but the error message would not be anymore terse than the compiler error(and there's no guarantee we could catch everything either since a lot of cppcheck's analysis is incomplete).

Reply:
Well I guess it's more productive for us if I choose a battle that matters more. Let's have this your way.

I think that templates can be useful in certain cases but it's best to be a bit cautious.

> Using const_cast this would not be caught.

I agree.. using const_cast will in theory leave the door open for UB.

Reply:
> I think that templates can be useful in certain cases but it's best to be a bit cautious.

Yes, of course, and I am weighing using the template against potential UB.

I am also trying to minimize its exposure. Its only added as an implementation detail in a cpp file that has concrete wrappers. So everyone should be calling the non-template function `nextAfterAstRightmostLeaf` and not the templated function.

I also added a check on the type to try and improve the error. I inserted a call to the template function using `Scope*` instead of `Token*` and this is the error I got from the compiler:

```
cppcheck/lib/astutils.cpp:300:55: error: no matching function for call to 'nextAfterAstRightmostLeafGeneric'
Token* test(Token* tok) { return nextAfterAstRightmostLeafGeneric(tok->scope()); }
                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cppcheck/lib/astutils.cpp:281:11: note: candidate template ignored: requirement 'std::is_convertible<const Scope *, const Token *>::value' was not satisfied [with T = const Scope]
static T* nextAfterAstRightmostLeafGeneric(T* tok)
          ^
```

First, the error points to the call to the function and not inside the template implementation. It also has a note about the constraint that failed(with the actual types substituted), so you can cleary see that `const Scope *` is not convertible to `const Token*`.

If you dont think these extra constraints are necessary(since these functions are implementation details), I can remove them, but I was trying to improve the error message since that is a major concern.

Reply:
That error message is a lot better. Let's keep that constraint.